confluentinc/cli
# [CLI-2356] Fix inability to hard delete schema versions

Release Notes
-------------


Breaking Changes
- PLACEHOLDER

New Features
- PLACEHOLDER

Bug Fixes
- Fix an issue where ``confluent schema-registry schema delete --permanent`` prevents permanent deletion of schema versions

Checklist
---------
1. [CRUCIAL] Is the change for features that are already live in prod?
   * yes: ok

What
----
Existence check fails for soft deleted schema versions without the `Deleted` optional argument in the `GET` request.

Unfortunately, using this optional parameter will still result in successful calls for non deleted schema versions in addition to soft deleted ones. So for the purposes of existence checking, we proceed when the `--permanent` flag is used if:

`GetSchemaByVersion` with `Deleted` returns no error and `GetSchemaByVersion` without `Deleted` returns an error.

I also updated the confirmation message to say "permanently delete" when that flag is supplied.

References
----------


Test & Review
-------------
Manual testing

Attempt to permanently delete a version that hasn't been soft deleted (expect failure):
```
confluent sr schema delete --subject test1 --version 1 --permanent
Error: you must first soft delete a schema version before you can hard delete it
```
Soft delete it (expect success):
```
confluent sr schema delete --subject test1 --version 1            
Are you sure you want to delete schema "test1 (version 1)"?
To confirm, type "test1". To cancel, press Ctrl-C: test1
Successfully soft deleted version "1" for subject "test1".
  Version  
-----------
        1  
```
Try to soft delete it again (expect failure):
```
confluent sr schema delete --subject test1 --version 1
Error: Schema Registry subject or version not found

Suggestions:
    List available subjects with `confluent schema-registry subject list`.
    List available versions with `confluent schema-registry subject describe`.
```
Now  hard delete it (expect success):
```
confluent sr schema delete --subject test1 --version 1 --permanent
Are you sure you want to permanently delete schema "test1 (version 1)"?
To confirm, type "test1". To cancel, press Ctrl-C: test1
Successfully hard deleted version "1" for subject "test1".
  Version  
-----------
        1  
```
Try to hard delete it again (expect failure):
```
confluent sr schema delete --subject test1 --version 1 --permanent
Error: Schema Registry subject or version not found

Suggestions:
    List available subjects with `confluent schema-registry subject list`.
    List available versions with `confluent schema-registry subject describe`.
```
Try to hard delete a version that doesn't exist (expect failure):
```
confluent sr schema delete --subject test1 --version 3 --permanent
Error: Schema Registry subject or version not found

Suggestions:
    List available subjects with `confluent schema-registry subject list`.
    List available versions with `confluent schema-registry subject describe`.
```
